lib/TPP/ConvertLinalgToXsmm.cpp

```diff
@@ -0,0 +1,338 @@
+  SmallVector<int64_t> strides;
+  int64_t offset;
+  if (failed(getStridesAndOffset(memref, strides, offset))) {
+    return failure();
+  }
+  if (llvm::any_of(strides, [](int64_t stride) {
+        return stride == ShapedType::kDynamic;
+      })) {
+    return failure();
+  }
+  if (strides.back() != 1)
+    return failure();
+  return strides;
+}
+
+static FailureOr<SmallVector<int64_t>> verifyStrides(OpOperand *operand) {
+  return verifyStrides(operand->get().getType());
+}
+
+// Return true if all the operand have the same type. No implicit conversion in
+// the linalgOp.
+static bool hasEqualTypes(linalg::LinalgOp linalgOp) {
+  // assert(linalgOp.getNumInitOperands > 0);
+  OpOperand *outputOperand = linalgOp.getDpsInitOperands().back();
+  auto elemType = getElementTypeOrSelf(outputOperand->get().getType());
+
+  if (!llvm::all_of(linalgOp.getDpsInitOperands(), [&](OpOperand *operand) {
+        auto currentOperandType =
+            getElementTypeOrSelf(operand->get().getType());
+        return currentOperandType == elemType;
+      })) {
+    return false;
+  }
+
+  return llvm::all_of(linalgOp.getDpsInputOperands(), [&](OpOperand *operand) {
+    auto currentOperandType = getElementTypeOrSelf(operand->get().getType());
+    return currentOperandType == elemType;
+  });
+}
+
+// Replace `linalgOp` with a unary dispatch plus invoke.
+static void replaceOpWithUnary(RewriterBase &rewriter,
+                               linalg::LinalgOp linalgOp, UnaryInfo unaryInfo,
+                               ArrayAttr flags, xsmm::UnaryKindAttr kind) {
+  Location loc = linalgOp.getLoc();
+  IntegerType integer64 = IntegerType::get(rewriter.getContext(), 64);
+  DenseI64ArrayAttr dims = DenseI64ArrayAttr::get(
+      rewriter.getContext(), ArrayRef<int64_t>{unaryInfo.m, unaryInfo.n,
+                                               unaryInfo.ldi, unaryInfo.ldo});
+  auto dtype = xsmm::utils::getDataType(
+      rewriter, linalgOp.getDpsInitOperands()[0]->get().getType());
+  Value dispatched = rewriter.create<xsmm::UnaryDispatchOp>(
+      loc, integer64, kind, dims, flags, dtype);
+  SmallVector<Value> invokeOperands;
+  invokeOperands.push_back(dispatched);
+  invokeOperands.append(linalgOp->getOperands().begin(),
+                        linalgOp->getOperands().end());
+  rewriter.replaceOpWithNewOp<xsmm::UnaryOp>(linalgOp, dtype, kind,
+                                             invokeOperands);
+}
+
+// Convert a linalg.fill to XSMM zero, if the fill fills with zeros.
+struct ConvertFillOpToUnaryZero : public OpRewritePattern<linalg::FillOp> {
+  using OpRewritePattern<linalg::FillOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::FillOp fillOp,
+                                PatternRewriter &rewriter) const override {
+    if (!fillOp.hasBufferSemantics() || fillOp.hasDynamicShape() ||
+        !hasEqualTypes(fillOp)) {
+      return failure();
+    }
+    auto input = fillOp.getDpsInputOperands()[0];
```

Comment:
Can we not use the matchers?

Reply:
Yes, we could use the matchers but it turned out to be more verbose here.

Reply:
The point of the matchers is not to reduce verbosity, but to make it clear what we're matching and be composable, so we can cascade them in a clear way. Going back to manually matching things makes it unclear and confusing.

Reply:
Nice matcher structure!